### ISSUE TITLE
CSS language server doesn't apply filtering

### ISSUE THREAD

**Issue by dbaeumer** [Microsoft/language-server-protocol#105]
_From @mickaelistria on October 18, 2016 8:51_

- VSCode Version: 1.5.3 (5be4091987a98e3870d89d630eb87be6d9bafd27)
- OS Version: Linux Fedora 24

I'm trying to consume the CSS language server from VSCode via stdio
`/usr/bin/node /usr/share/code/resources/app/extensions/css/server/out/cssServerMain.js --stdio`
and connect to it using the https://github.com/eclipselabs/eclipse-language-service and underlying https://github.com/TypeFox/ls-api
* Connection works, request are well sent and result well received.
* When trying to complete after the `a` in
```css
a {
}
```
I see that `:checked` is proposed, whereas it doesn't make sense here. If I try VSCode directly, the `:checked` isn't proposed. It seems like VSCode has some additional filtering for the CSS pseudo-classes, and that this filtering is not part of the language server.

Could this filtering be part on the language server side?

_Copied from original issue: Microsoft/vscode#13926_

**Comment by dbaeumer** [comment]
_From @aeschli on October 19, 2016 7:22_

Yes, that's how it is designed. Filtering has to be done client side. Some clients want to only filter by prefix, some using camel-case or fuzzy-matching. For consistency across languages we let the client do the filtering.
Assigning to @dbaeumer to add that to the language server protocol documentation.

**Comment by dbaeumer** [comment]
_From @mickaelistria on October 19, 2016 7:27_

So that means that the client has to do some level of file-reading to at least guess the word it applies to? I have the impression this is contrary to the goals of the protocol. Maybe the filtering strategy should be passed by clients as an option when starting the server, so the server would apply the requested strategy. Default strategy would be "none" (like now), other possible value would be "fuzzy" or "prefix".

**Comment by kaloyan-raev** [comment]
If we agree that the server is responsible for the filtering then we should add another property to the `CompletionItem` - `highlights`. The latter would mark which parts of the label should be highlighted in the client's UI based on the user input.

**Comment by dbaeumer** [comment]
Standardizing on the filter will be very hard since tools usually come with a filtering algorithm. And a goal of the tool is to make this the same for different languages. So I opt to specify this as follows:

- clients can request "prefix" or "none".
- servers tell clients if they honored the client request.

For maximal tool flexibility it is recommend to implement none and to do the filtering on the client side. If we hit performance problems with this then I think we should look into pagination instead of doing filtering on the server side.

**Comment by mickaelistria** [comment]
> ools usually come with a filtering algorithm.

Many tools also come with a completion engine, and it doesn't prevent the LSP to provide completion ;)
The underlying question is whether computing the conextual "word to complete" is something that should be part of the language server or the tool. If you consider the most exotic languages such as Brainfuck or WhiteSpace, then finding a "word" would require clients to set up language-specific logic. It seems to me that it's somehow what we want to avoid with a LSP.

In the context of the CSS server from VSCode, let just say that completing `:checked` against the `a` object type seems to be a bug, and that this bug is IMHO caused by missing filtering on results.

**Comment by aeschli** [comment]
Agree, 'word to complete' is another issue we need to look at. I agree that we should avoid it and think we remove it from a future version of the protocol. Best is if servers always returning text edits in completion proposals. That's what currently the vscode css, json and html language servers all do.
We should add that as advice to the current version.
